File: blind75/42_reorder_list/reorder_list.py
from typing import Optional


class ListNode:
    """Definition for singly-linked list node."""

    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next


def reorder_list(head: Optional[ListNode]) -> None:
    """
    Reorders the list in-place using find middle + reverse + merge.
    Time complexity: O(n), Space complexity: O(1)

    Args:
        head: The head of the linked list

    Note:
        Modifies the list in-place. Does not return anything.
    """
    if head is None or head.next is None:
        return

    # Step 1: Find the middle of the list
    slow, fast = head, head
    while fast.next is not None and fast.next.next is not None:
        slow = slow.next
        fast = fast.next.next

    # Step 2: Split and reverse the second half
    second = slow.next
    slow.next = None  # Split the list
    second = reverse_list(second)

    # Step 3: Merge the two halves
    first = head
    while second is not None:
        tmp1 = first.next
        tmp2 = second.next

        first.next = second
        second.next = tmp1

        first = tmp1
        second = tmp2


def reverse_list(head: Optional[ListNode]) -> Optional[ListNode]:
    """Helper function to reverse a linked list."""
    prev = None
    curr = head

    while curr is not None:
        next_node = curr.next
        curr.next = prev
        prev = curr
        curr = next_node

    return prev


def reorder_list_stack(head: Optional[ListNode]) -> None:
    """
    Reorders the list using a stack.
    Time complexity: O(n), Space complexity: O(n)

    Args:
        head: The head of the linked list

    Note:
        Modifies the list in-place. Does not return anything.
    """
    if head is None or head.next is None:
        return

    # Push all nodes to stack
    stack = []
    curr = head
    while curr is not None:
        stack.append(curr)
        curr = curr.next

    # Reorder by taking from front and back alternately
    curr = head
    n = len(stack)
    for i in range(n // 2):
        next_node = curr.next
        back = stack.pop()

        curr.next = back
        back.next = next_node

        curr = next_node

    # Set the end of the list
    if curr is not None:
        curr.next = None

File: blind75/42_reorder_list/reorder_list.go
package reorder_list

// ListNode represents a node in a singly linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

// ReorderList reorders the list in-place using find middle + reverse + merge.
// Time complexity: O(n), Space complexity: O(1)
func ReorderList(head *ListNode) {
	if head == nil || head.Next == nil {
		return
	}

	// Step 1: Find the middle of the list
	slow, fast := head, head
	for fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	// Step 2: Split and reverse the second half
	second := slow.Next
	slow.Next = nil // Split the list
	second = reverseList(second)

	// Step 3: Merge the two halves
	first := head
	for second != nil {
		tmp1 := first.Next
		tmp2 := second.Next

		first.Next = second
		second.Next = tmp1

		first = tmp1
		second = tmp2
	}
}

// reverseList reverses a linked list
func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	curr := head

	for curr != nil {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}

	return prev
}

// ReorderListStack reorders the list using a stack.
// Time complexity: O(n), Space complexity: O(n)
func ReorderListStack(head *ListNode) {
	if head == nil || head.Next == nil {
		return
	}

	// Push all nodes to stack
	var stack []*ListNode
	curr := head
	for curr != nil {
		stack = append(stack, curr)
		curr = curr.Next
	}

	// Reorder by taking from front and back alternately
	curr = head
	n := len(stack)
	for i := 0; i < n/2; i++ {
		next := curr.Next
		back := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		curr.Next = back
		back.Next = next

		curr = next
	}

	// Set the end of the list
	if curr != nil {
		curr.Next = nil
	}
}
